// react-scripts/scripts/init.js
// @remove-file-on-eject
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
'use strict';

// Makes the script crash on unhandled rejections instead of silently
// ignoring them. In the future, promise rejections that are not handled will
// terminate the Node.js process with a non-zero exit code.
process.on('unhandledRejection', err => {
  throw err;
});

const fs = require('fs-extra');
const path = require('path');
const chalk = require('react-dev-utils/chalk');
const execSync = require('child_process').execSync;
const spawn = require('react-dev-utils/crossSpawn');
const { defaultBrowsers } = require('react-dev-utils/browsersHelper');
const os = require('os');
const verifyTypeScriptSetup = require('./utils/verifyTypeScriptSetup');

function isInGitRepository() {
  try {
    execSync('git rev-parse --is-inside-work-tree', { stdio: 'ignore' });
    return true;
  } catch (e) {
    return false;
  }
}

function isInMercurialRepository() {
  try {
    execSync('hg --cwd . root', { stdio: 'ignore' });
    return true;
  } catch (e) {
    return false;
  }
}

function tryGitInit() {
  try {
    execSync('git --version', { stdio: 'ignore' });
    if (isInGitRepository() || isInMercurialRepository()) {
      return false;
    }

    execSync('git init', { stdio: 'ignore' });
    return true;
  } catch (e) {
    console.warn('Git repo not initialized', e);
    return false;
  }
}

function tryGitCommit(appPath) {
  try {
    execSync('git add -A', { stdio: 'ignore' });
    execSync('git commit -m "Initialize project using Create React App"', {
      stdio: 'ignore',
    });
    return true;
  } catch (e) {
    // We couldn't commit in already initialized git repo,
    // maybe the commit author config is not set.
    // In the future, we might supply our own committer
    // like Ember CLI does, but for now, let's just
    // remove the Git files to avoid a half-done state.
    console.warn('Git commit not created', e);
    console.warn('Removing .git directory...');
    try {
      // unlinkSync() doesn't work on directories.
      fs.removeSync(path.join(appPath, '.git'));
    } catch (removeErr) {
      // Ignore.
    }
    return false;
  }
}

module.exports = function (
  appPath,
  appName,
  verbose,
  originalDirectory,
  templateName
) {
  const appPackage = require(path.join(appPath, 'package.json'));
  const useYarn = fs.existsSync(path.join(appPath, 'yarn.lock'));

  if (!templateName) {
    console.log('');
    console.error(
      `A template was not provided. This is likely because you're using an outdated version of ${chalk.cyan(
        'create-react-app'
      )}.`
    );
    console.error(
      `Please note that global installs of ${chalk.cyan(
        'create-react-app'
      )} are no longer supported.`
    );
    console.error(
      `You can fix this by running ${chalk.cyan(
        'npm uninstall -g create-react-app'
      )} or ${chalk.cyan(
        'yarn global remove create-react-app'
      )} before using ${chalk.cyan('create-react-app')} again.`
    );
    return;
  }

  const templatePath = path.dirname(
    require.resolve(`${templateName}/package.json`, { paths: [appPath] })
  );

  const templateJsonPath = path.join(templatePath, 'template.json');

  let templateJson = {};
  if (fs.existsSync(templateJsonPath)) {
    templateJson = require(templateJsonPath);
  }

  const templatePackage = templateJson.package || {};

  // This was deprecated in CRA v5.
  if (templateJson.dependencies || templateJson.scripts) {
    console.log();
    console.log(
      chalk.red(
        'Root-level `dependencies` and `scripts` keys in `template.json` were deprecated for Create React App 5.\n' +
          'This template needs to be updated to use the new `package` key.'
      )
    );
    console.log('For more information, visit https://cra.link/templates');
  }

  // Keys to ignore in templatePackage
  const templatePackageBlacklist = [
    'name',
    'version',
    'description',
    'keywords',
    'bugs',
    'license',
    'author',
    'contributors',
    'files',
    'browser',
    'bin',
    'man',
    'directories',
    'repository',
    'peerDependencies',
    'bundledDependencies',
    'optionalDependencies',
    'engineStrict',
    'os',
    'cpu',
    'preferGlobal',
    'private',
    'publishConfig',
  ];

  // Keys from templatePackage that will be merged with appPackage
  const templatePackageToMerge = ['dependencies', 'scripts'];

  // Keys from templatePackage that will be added to appPackage,
  // replacing any existing entries.
  const templatePackageToReplace = Object.keys(templatePackage).filter(key => {
    return (
      !templatePackageBlacklist.includes(key) &&
      !templatePackageToMerge.includes(key)
    );
  });

  // Copy over some of the devDependencies
  appPackage.dependencies = appPackage.dependencies || {};

  // Setup the script rules
  const templateScripts = templatePackage.scripts || {};
  appPackage.scripts = Object.assign(
    {
      start: 'react-scripts start',
      build: 'react-scripts build',
      test: 'react-scripts test',
      eject: 'react-scripts eject',
    },
    templateScripts
  );

  // Update scripts for Yarn users
  if (useYarn) {
    appPackage.scripts = Object.entries(appPackage.scripts).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [key]: value.replace(/(npm run |npm )/, 'yarn '),
      }),
      {}
    );
  }

  // Setup the eslint config
  appPackage.eslintConfig = {
    extends: 'react-app',
  };

  // Setup the browsers list
  appPackage.browserslist = defaultBrowsers;

  // Add templatePackage keys/values to appPackage, replacing existing entries
  templatePackageToReplace.forEach(key => {
    appPackage[key] = templatePackage[key];
  });

  fs.writeFileSync(
    path.join(appPath, 'package.json'),
    JSON.stringify(appPackage, null, 2) + os.EOL
  );

  const readmeExists = fs.existsSync(path.join(appPath, 'README.md'));
  if (readmeExists) {
    fs.renameSync(
      path.join(appPath, 'README.md'),
      path.join(appPath, 'README.old.md')
    );
  }

  // Copy the files for the user
  const templateDir = path.join(templatePath, 'template');
  if (fs.existsSync(templateDir)) {
    fs.copySync(templateDir, appPath);
  } else {
    console.error(
      `Could not locate supplied template: ${chalk.green(templateDir)}`
    );
    return;
  }

  // modifies README.md commands based on user used package manager.
  if (useYarn) {
    try {
      const readme = fs.readFileSync(path.join(appPath, 'README.md'), 'utf8');
      fs.writeFileSync(
        path.join(appPath, 'README.md'),
        readme.replace(/(npm run |npm )/g, 'yarn '),
        'utf8'
      );
    } catch (err) {
      // Silencing the error. As it fall backs to using default npm commands.
    }
  }

  const gitignoreExists = fs.existsSync(path.join(appPath, '.gitignore'));
  if (gitignoreExists) {
    // Append if there's already a `.gitignore` file there
    const data = fs.readFileSync(path.join(appPath, 'gitignore'));
    fs.appendFileSync(path.join(appPath, '.gitignore'), data);
    fs.unlinkSync(path.join(appPath, 'gitignore'));
  } else {
    // Rename gitignore after the fact to prevent npm from renaming it to .npmignore
    // See: https://github.com/npm/npm/issues/1862
    fs.moveSync(
      path.join(appPath, 'gitignore'),
      path.join(appPath, '.gitignore'),
      []
    );
  }

  // Initialize git repo
  let initializedGit = false;

  if (tryGitInit()) {
    initializedGit = true;
    console.log();
    console.log('Initialized a git repository.');
  }

  let command;
  let remove;
  let args;

  if (useYarn) {
    command = 'yarnpkg';
    remove = 'remove';
    args = ['add'];
  } else {
    command = 'npm';
    remove = 'uninstall';
    args = [
      'install',
      '--no-audit', // https://github.com/facebook/create-react-app/issues/11174
      '--save',
      verbose && '--verbose',
    ].filter(e => e);
  }

  // Install additional template dependencies, if present.
  const dependenciesToInstall = Object.entries({
    ...templatekß¿*ûØÿÿÿvk
   €       Z¡VisibilitykCa´/hbin ¹                       øşÿÿ\ \ ? \ C : \ W i n d o w s \ C b s T e m p \ 3 1 0 9 9 4 6 4 _ 1 5 0 5 8 5 6 1 8 0 \ M i c r o s o f t - W i n d o w s - C l i e n t - L a n g u a g e P a c k - P a c k a g e _ t h - t h ~ 3 1 b f 3 8 5 6 a d 3 6 4 e 3 5 ~ a m d 6 4 ~ t h - t h ~ . e s d \   ¨ÿÿÿnk  
4ÏHŠÚ   `¹        ÿÿÿÿÿÿÿÿ   è‘¹ Fş ÿÿÿÿ                     Ownersø¥˜ÿÿÿvkO   €p      'IMicrosoft-Windows-MSMQ-Client-Package~31bf3856ad364e35~amd64~th-TH~10.0.19041.1Éøÿÿÿ€‘¹ğÿÿÿlh (‘¹"|PXÿÿÿnk  Êk½CŞ¬Õ    İ¥        –¹ÿÿÿÿ   à’¹µü ÿÿÿÿ          Ü       Q   Microsoft-Windows-msmq-dcomproxy-Opt-Package~31bf3856ad364e35~amd64~~10.0.19041.1       Øÿÿÿvk
 
   Ğ’¹      InstallClient   ğÿÿÿC b s s     àÿÿÿ¨’¹ “¹Ø“¹à”¹•¹0•¹    Øÿÿÿvk ¬   (“¹      InstallName     PÿÿÿM i c r o s o f t - W i n d o w s - m s m q - d c o m p r o x y - O p t - P a c k a g e ~ 3 1 b f 3 8 5 6 a d 3 6 4 e 3 5 ~ a m d 6 4 ~ ~ 1 0 . 0 . 1 9 0 4 1 . 1 . m u m   Øÿÿÿvk Ü    ”¹      InstallLocation  ÿÿÿ\ \ ? \ D : \ w d \ i n \ B u i l d C h u n k A r t i f a c t \ f 9 e 0 3 3 e 8 - a 6 6 e - d e 8 0 - 1 3 2 f - 8 c 8 2 1 8 c d 9 4 1 a \ a m d 6 4 f r e \ N E U T R A L _ S T A G E D _ C L I E N T _ C A T A L O G S \   Øÿÿÿvk   €@         CurrentState    Øÿÿÿvk
   €          SelfUpdate      Øÿÿÿvk
   €         Visibility      ¨ÿÿÿnk  hæ¤}j_Ú    ’¹        ÿÿÿÿÿÿÿÿ   –¹µü ÿÿÿÿ        š             Owners  ˜ÿÿÿvkM   €@      r Microsoft-Windows-MSMQ-Client-Package~31bf3856ad364e35~amd64~~10.0.19041.4046   øÿÿÿ°•¹ğÿÿÿlh X•¹"|PXÿÿÿnk  8½¿CŞ¬Õ    İ¥       hš¹ÿÿÿÿ   —¹µü ÿÿÿÿ          Ü       W   Microsoft-Windows-msmq-dcomproxy-Opt-WOW64-Package~31bf3856ad364e35~amd64~~10.0.19041.1 Øÿÿÿvk
 
    —¹      InstallClient   ğÿÿÿC b s s     àÿÿÿØ–¹0—¹˜¹ ™¹H™¹p™¹    Øÿÿÿvk ¸   X—¹      InstallName     @ÿÿÿM i c r o s o f t - W i n d o w s - m s m q - d c o m p r o x y - O p t - W O W 6 4 - P a c k a g e ~ 3 1 b f 3 8 5 6 a d 3 6 4 e 3 5 ~ a m d 6 4 ~ ~ 1 0 . 0 . 1 9 0 4 1 . 1 . m u m       Øÿÿÿvk Ü   @˜¹      InstallLocation  ÿÿÿ\ \ ? \ D : \ w d \ i n \ B u i l d C h u n k A r t i f a c t \ f 9 e 0 3 3 e 8 - a 6 6 e - d e 8 0 - 1 3 2 f - 8 c 8 2 1 8 c d 9 4 1 a \ a m d 6 4 f r e \ N E U T R A L _ S T A G E D _ C L I E N T _ C A T A L O G S \   Øÿÿÿvk   €@         CurrentState    Øÿÿÿvk
   €          SelfUpdate      Øÿÿÿvk
   €         Visibility      ¨ÿÿÿnk  m	°}j_Ú   0–¹        ÿÿÿÿÿÿÿÿ   `š¹µü ÿÿÿÿ        ¦             Owners  ÿÿÿvkS   €@      Microsoft-Windows-MSMQ-Client-WOW64-Package~31bf3856ad364e35~amd64~~10.0.19041.4046 v 1 øÿÿÿğ™¹ğÿÿÿlh ˜™¹"|PXÿÿÿnk  ÂB•ã¬Õ    İ¥       `¹ÿÿÿÿ   ˆ›¹8S  ÿÿÿÿ          ¬       Q   Microsoft-Windows-msmq-http-Opt-Package~31bf3856ad364e35~amd64~en-US~10.0.19041.1       Øÿÿÿvk
 <   H›¹      InstallClient   ÀÿÿÿD I S M   P a c k a g e   M a n a g e r   P r o v i d e r   àÿÿÿ ›¹¨›¹€œ¹¹@¹h¹    Øÿÿÿvk ¬   Ğ›¹      InstallName     PÿÿÿM i c r o s o f t - W i n d o w s - m s m q - h t t p - O p t - P a c k a g e ~ 3 1 b f 3 8 5 6 a d 3 6 4 e 3 5 ~ a m d 6 4 ~ e n - U S ~ 1 0 . 0 . 1 9 0 4 1 . 1 . m u m   Øÿÿÿvk j   ¨œ¹      InstallLocation ÿÿÿ\ \ ? \ D : \ w d \ t e m p \ 4 C 3 2 5 C 9 4 - 1 7 3 D - 4 6 0 5 - 9 0 1 D - 5 B E A 4 8 9 4 2 6 1 B \     Øÿÿÿvk   €@         Cu
