// jsdom/lib/jsdom/living/generated/HTMLSelectElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLConstructor_helpers_html_constructor = require("../helpers/html-constructor.js").HTMLConstructor;
const HTMLOptionElement = require("./HTMLOptionElement.js");
const HTMLOptGroupElement = require("./HTMLOptGroupElement.js");
const HTMLElement = require("./HTMLElement.js");
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const parseNonNegativeInteger_helpers_strings = require("../helpers/strings.js").parseNonNegativeInteger;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;

const interfaceName = "HTMLSelectElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
};
exports.convert = (value, { context = "The provided value" } = {}) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }
  throw new TypeError(`${context} is not of type 'HTMLSelectElement'.`);
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  const ctor = globalObject[ctorRegistrySymbol]["HTMLSelectElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLSelectElement is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

function makeProxy(wrapper, globalObject) {
  let proxyHandler = proxyHandlerCache.get(globalObject);
  if (proxyHandler === undefined) {
    proxyHandler = new ProxyHandler(globalObject);
    proxyHandlerCache.set(globalObject, proxyHandler);
  }
  return new Proxy(wrapper, proxyHandler);
}

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = (wrapper, globalObject) => {
  HTMLElement._internalSetup(wrapper, globalObject);
};

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {}) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });

  wrapper = makeProxy(wrapper, globalObject);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }
  return wrapper;
};

exports.new = globalObject => {
  let wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });

  wrapper = makeProxy(wrapper, globalObject);

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }
  return wrapper[implSymbol];
};

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  }

  if (globalObject.HTMLElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLSelectElement before HTMLElement");
  }
  class HTMLSelectElement extends globalObject.HTMLElement {
    constructor() {
      return HTMLConstructor_helpers_html_constructor(globalObject, interfaceName, new.target);
    }

    item(index) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'item' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'item' on 'HTMLSelectElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      }
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'item' on 'HTMLSelectElement': parameter 1"
        });
        args.push(curArg);
      }
      return utils.tryWrapperForImpl(esValue[implSymbol].item(...args));
    }

    namedItem(name) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'namedItem' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'namedItem' on 'HTMLSelectElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      }
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'namedItem' on 'HTMLSelectElement': parameter 1"
        });
        args.push(curArg);
      }
      return utils.tryWrapperForImpl(esValue[implSymbol].namedItem(...args));
    }

    add(element) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'add' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'add' on 'HTMLSelectElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      }
      const args = [];
      {
        let curArg = arguments[0];
        if (HTMLOptionElement.is(curArg) || HTMLOptGroupElement.is(curArg)) {
          curArg = utils.implForWrapper(curArg);
        } else {
          throw new TypeError(
            "Failed to execute 'add' on 'HTMLSelectElement': parameter 1" + " is not of any supported type."
          );
        }
        args.push(curArg);
      }
      {
        let curArg = arguments[1];
        if (curArg !== undefined) {
          if (curArg === null || curArg === undefined) {
            curArg = null;
          } else {
            if (HTMLElement.is(curArg)) {
              curArg = utils.implForWrapper(curArg);
            } else if (typeof curArg === "number") {
              curArg = conversions["long"](curArg, {
                context: "Failed to execute 'add' on 'HTMLSelectElement': parameter 2"
              });
            } else {
              curArg = conversions["long"](curArg, {
                context: "Failed to execute 'add' on 'HTMLSelectElement': parameter 2"
              });
            }
          }
        } else {
          curArg = null;
        }
        args.push(curArg);
      }
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].add(...args);
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    remove() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'remove' called on an object that is not a valid instance of HTMLSelectElement.");
      }
      const args = [];
      switch (arguments.length) {
        case 0:
          break;
        default: {
          let curArg = arguments[0];
          curArg = conversions["long"](curArg, {
            context: "Failed to execute 'remove' on 'HTMLSelectElement': parameter 1"
          });
          args.push(curArg);
        }
      }
      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].remove(...args);
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    checkValidity() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'checkValidity' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      return esValue[implSymbol].checkValidity();
    }

    reportValidity() {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError("'reportValidity' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      return esValue[implSymbol].reportValidity();
    }

    setCustomValidity(error) {
      const esValue = this !== null && this !== undefined ? this : globalObject;
      if (!exports.is(esValue)) {
        throw new TypeError(
          "'setCustomValidity' called on an object that is not a valid instance of HTMLSelectElement."
        );
      }

      if (arguments.length < 1) {
        throw new TypeError(
          "Failed to execute 'setCustomValidity' on 'HTMLSelectElement': 1 argument required, but only " +
            arguments.length +
            " present."
        );
      }
      const args = [];
      {
        let curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'setCustomValidity' on 'HTMLSelectElement': parameter 1"
        });
        args.push(curArg);
      }
      return esValue[implSymbol].setCustomValidity(...args);
    }

    get autofocus() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get autofocus' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "autofocus");
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    set autofocus(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set autofocus' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'autofocus' property on 'HTMLSelectElement': The provided value"
      });

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "autofocus", "");
        } else {
          esValue[implSymbol].removeAttributeNS(null, "autofocus");
        }
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    get disabled() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get disabled' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        return esValue[implSymbol].hasAttributeNS(null, "disabled");
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    set disabled(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set disabled' called on an object that is not a valid instance of HTMLSelectElement.");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'disabled' property on 'HTMLSelectElement': The provided value"
      });

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        if (V) {
          esValue[implSymbol].setAttributeNS(null, "disabled", "");
        } else {
   Øÿÿÿvk N   Øğ¦      HardwareKeyboardøÿÿÿèªªhbin ğ¦                       ¨ÿÿÿ{ E 0 C 1 5 D 5 7 - C 7 1 7 - 4 4 b 6 - A 1 9 0 - 1 C 8 4 2 E 3 6 0 7 B 0 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   0ñ¦ `= ÿÿÿÿ            N       	   LANG_0C09       ¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿĞï¦@ñ¦    Øÿÿÿvk N   hñ¦      SoftKeyboard    ¨ÿÿÿ{ E 0 C 1 5 D 5 7 - C 7 1 7 - 4 4 b 6 - A 1 9 0 - 1 C 8 4 2 E 3 6 0 7 B 0 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ    ò¦ `= ÿÿÿÿ            N       	   LANG_0C0A       Øÿÿÿvk N   Hò¦      HardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿ ò¦°ò¦    Øÿÿÿvk N   Øò¦      SoftKeyboard    ¨ÿÿÿ{ 3 d f e 5 d d e - 9 4 f 9 - 4 e 8 7 - 9 1 d 0 - 1 6 7 d f 4 8 d b a 8 8 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   ô¦ `= ÿÿÿÿ            N       	   LANG_0C0C       Øÿÿÿvk N   ¸ó¦      HardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿó¦ ô¦    Øÿÿÿvk N   Hô¦      SoftKeyboard    ¨ÿÿÿ{ 3 1 C 0 4 C 2 6 - 5 D 2 8 - 4 9 1 3 - A 7 D F - 0 1 2 2 B 4 4 9 2 7 D 1 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   €õ¦ `= ÿÿÿÿ            N       	   LANG_1009       Øÿÿÿvk N   (õ¦      HardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿ õ¦õ¦    Øÿÿÿvk N   ¸õ¦      SoftKeyboard    ¨ÿÿÿ{ E 0 C 1 5 D 5 7 - C 7 1 7 - 4 4 b 6 - A 1 9 0 - 1 C 8 4 2 E 3 6 0 7 B 0 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   ğö¦ `= ÿÿÿÿ            N       	   LANG_100C       Øÿÿÿvk N   ˜ö¦      HardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿpö¦ ÷¦    Øÿÿÿvk N   (÷¦      SoftKeyboard    ¨ÿÿÿ{ 0 6 e 8 a 9 1 f - 3 5 e 3 - 4 2 8 d - a 7 2 2 - b e 4 5 b d 1 9 5 7 8 d }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   `ø¦ `= ÿÿÿÿ            N       	   LANG_141A       Øÿÿÿvk N   ø¦      HardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿà÷¦pø¦    Øÿÿÿvk N   ˜ø¦      SoftKeyboard    ¨ÿÿÿ{ F C 4 4 5 6 5 7 - E D 0 8 - 1 1 e 3 - A 6 9 E - B 4 B 5 2 F E 0 6 6 1 1 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   Ğù¦ `= ÿÿÿÿ            N       	   LANG_1809       Øÿÿÿvk N   xù¦    —ñHardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿPù¦àù¦    Øÿÿÿvk N   ú¦      SoftKeyboard    ¨ÿÿÿ{ C 2 E E B 5 E F - D 9 D 7 - 4 D 8 C - 8 7 9 D - 5 B 7 A C 7 B 5 B F 6 5 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   @û¦ `= ÿÿÿÿ            N       	   LANG_241A       Øÿÿÿvk N   èú¦    MBHardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿÀú¦Pû¦    Øÿÿÿvk N   xû¦    MBSoftKeyboard    ¨ÿÿÿ{ 5 8 4 4 5 6 5 7 - 6 0 7 D - 1 1 e 4 - 9 9 B 9 - B 4 B 5 2 F E 0 6 6 1 1 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ   °ü¦ `= ÿÿÿÿ            N       	   LANG_281A       Øÿÿÿvk N   Xü¦    —ñHardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿ0ü¦Àü¦    Øÿÿÿvk N   èü¦    MBSoftKeyboard    ¨ÿÿÿ{ 6 4 4 4 5 6 5 7 - 6 6 E D - 1 1 e 4 - 9 9 B B - B 4 B 5 2 F E 0 6 6 1 1 }          ÿÿÿnk  ô$—ÕŞ¬Õ   (øG         ÿÿÿÿÿÿÿÿ    ş¦ `= ÿÿÿÿ            N       	   LANG_4009       Øÿÿÿvk N   Èı¦    MBHardwareKeyboard¨ÿÿÿ{ 8 0 4 c 7 2 d 2 - b 3 a 3 - 4 f 9 6 - a 5 6 8 - e b 8 7 7 5 e b 4 b 9 7 }         ğÿÿÿ ı¦0ş¦    Øÿÿÿvk N   Xş¦    —ñSoftKeyboard    ¨ÿÿÿ{ E 0 C 1 5 D 5 7 - C 7 1 7 - 4 4 b 6 - A 1 9 0 - 1 C 8 4 2 E 3 6 0 7 B 0 }         ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   Pÿ¦ `= ÿÿÿÿ        (          &   {06e8a91f-35e3-428d-a722-be45bd19578d}  Øÿÿÿvk ´     §      AccentTable     èÿÿÿ(ÿ¦hÿ¦˜ÿ¦Èÿ¦à§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk    Ø §      KeyTableğÿÿÿ' . -       øÿÿÿh«ªhbin  §                       Hÿÿÿd e d é d è d ê d ë     d u d ù d û d ü     d i d ï d î     d o d ô d S    d a d à d â d æ     d s d ß d §     d y d ÿ     d c d ç d ©     d m d µ     d . d , d ? d : d ! d -     øüÿÿ  q %w a       w .q <e –a %s       e ¾w †r as d       r –e €t Œd ;f       t Œr ˆz qf 0g       z |t _u Fg h       u 7z &i h ÿj       i Şu no j @k       o _i p õk Al       p o l       a wq 0w Qs ºy       s )w Qe |a ]d >y ­x       d ^e r –s f ‘y Œx >c       f ²r ½t ¢d œg 0x ªc v       g t ‘z ‡f šh c kv ”b       h „z Xu ğg ]j >v äb n       j èu Yi Wh k Yb vn 0m       k i Òo ‚j ºl ‚n Im       l ¯o ap pk ¨m       y Âa ås bd x       x Òs ôd èf Cy Xc       c ôd Ïf ìg Nx §v       v ½f Og ğh ™c ‚b       b äg äh j `v ±n       n ùh ôj k b üm       m Öj Ék 0l n         c ôv b dn       , èb èn èm       . èn èm             Øÿÿÿvk    èÿ¦      SymbolsToInsert ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   H§ `= ÿÿÿÿ        (   8      &   {075f7e2b-3c34-4c3a-9312-9cafacfa31e6}  Øÿÿÿvk ˜   ¨§      AccentTable     `ÿÿÿd e d è d é     d u d ú d ü     d i d í d ï     d o d ò d ó     d a d à     d s d §     d l d @    d c d ç d ©     d m d µ     d . d ? d : d ! d -         èÿÿÿ€§`§§À§ 	§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk 8  à§      KeyTableÀüÿÿ    ƒc av Cb Yn       q Ww Ea Ğs       w í q Ce 3a ps Åd       e ü w :r  s ‰d 1f       r e t d êf dg       t r  y 8f Òg Âh       y t u  g »h ¡j       u Oy ù i h ¾j †k       i Vu Ø o j âk Ul       o `i Ï p ¢k Æl >ñ       p üo l  ñ       a zq °w *s z       s lq Îw ëe Ô a d ïz Îx       d vw ìe ‰r "s f z x Zc       f ’e $r $d ,g ßx $c ’v       g Rr —t Õy .f h ÷c Èv       h ‹t ‹y u Èg j Fv Øb ‹n       j y i ‡h í k b Ìn m       k u Ği ¨j Å l ‹m       l Gi Öo ˜p ºk · ñ       ñ o (p ‡l ãm       z Ê a [s ¿d ;x       x ms md mf ­ z c       c Qd €f !g Ş x v       v c ı b       b ªg ªh ªj ñ v Ú n       n õh :j õk sb ¿ m       m fj Ôk  l 3ñ cn       ,               .             Øÿÿÿvk    H	§      SymbolsToInsert ğÿÿÿ' . -       ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   8
§ `= ÿÿÿÿ        (          &   {15517764-3384-441f-a33c-b682776ddae4}  Øÿÿÿvk 8   ø	§      AccentTable     Àÿÿÿd e d é d ë     d i d ï     d o d ó d ö     d u d ü         èÿÿÿĞ	§P
§€
§°
§Ø
§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk    Ğ
§      KeyTableøüÿÿ  q Mw "a       w 
q je Ua Fs       e ¨w £r Ms ?d       r ‰e t ˆd zf       t ™r „y –f ag       y ®t _u ‚g Fh       u y Mi }h Xj       i Úu o .j Xk       o ui øp k pl       p ˆo ıl       a wq w Ès Oz       s Ow ne na ¬d az )x       d ue ôr s œf wz [x ıc       f År èt …d Æg Fx c èv       g t „y ~f Èh )c pv b       h ay Fu Ög j wv Jb ùn       j #u ôi jh ¾k Ob n Xm       k i o nj l ¯n #m       l Öo Fp •k äm       z Za Ùs ıd :x       x †s ıd èf xz Uc       c İd f ag {x ¶v       v #f 0g wh ;c §b       b g h Xj bv Ín       n ah j Fk mb ím       m „j Ék l dn         0c >v {b xn       , èb èn èm       . èn èm             Øÿÿÿvk     §      SymbolsToInsert ğÿÿÿ' . -       ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   (§ `= ÿÿÿÿ        (   ô      &   {248df235-c484-4e71-a48c-0340b61df858}  Øÿÿÿvk p   °§      AccentTable     ˆÿÿÿd ±d ¬    d µd ­    d Ãd Â    d ¹d ¯d Êd     d ¿d Ì    d Åd Íd Ëd °    d ·d ®    d Éd Î        èÿÿÿˆ§@§p§ §À§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk ô   §      KeyTableØÿÿÿvk    è§      SymbolsToInsert ğÿÿÿ' . -       øÿÿÿÀ­ªhbin §                        ıÿÿ  ; gÂ±wÃ      Â­; Sµ­±­Ãè´      µoÂ´ÁqÃ´ùÆ      Áwµ˜Ä†´,Æı³      ÄyÁ|Å²Æ-³ı·      ÅwÄ…¸)³@·)¾      ¸ûÅ>¹·¾Éº      ¹¯¸y¿ì¾ºä»      ¿c¹æÀZºû»      ÀP¿˜»      ±l; aÂHÃ      Ãè; „Â´µË±A´B¶      ´èÂ²µÁ¸Ã@Æì¶Ç      ÆµÖÁÄ¢´o³ôÇšÈ      ³ªÁàÄXÅ«ÆM·ºÈºÉ      ·èÄÅ¥¸b³*¾É´²      ¾èÅÅ¸Å¹ó·1ºÙ²Y½      ºè¸è¹è¿¾å »½¼      »¡¹½¿èÀáºJ¼      ¶äÃº´KÇ      Ç)´„ÆI¶?È      ÈÆ³lÇ‚É      É0³ì·0È¦²      ²„·¾7É·½      ½¾½º²ñ¼      ¼­º„»+½        ÉÈºÉ.²½      . è½è¼        ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   0§ `= ÿÿÿÿ        (          &   {29B0C7B8-7BB1-4ED9-8313-94F5F15C9B28}  Øÿÿÿvk t   ¸§      AccentTable     ˆÿÿÿd e d é     d u d ü     d o d ö     d a d ä     d s d ß d §     d c d ©     d m d µ     d . d , d ? d : d ! d -     èÿÿÿ§H§x§¨§Ğ§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk    È§      KeyTableøüÿÿ  q %w a       w .q <e –a %s       e ¾w †r as d       r –e €t Œd ;f       t Œr ˆz qf 0g       z |t _u Fg h       u 7z &i h ÿj       i Şu no j @k       o _i p õk Al       p o l       a wq 0w Qs ºy       s )w Qe |a ]d >y ­x       d ^e r –s f ‘y Œx >c       f ²r ½t ¢d œg 0x ªc v       g t ‘z ‡f šh c kv ”b       h „z Xu ğg ]j >v äb n       j èu Yi Wh k Yb vn 0m       k i Òo ‚j ºl ‚n Im       l ¯o ap pk ¨m       y Âa ås bd x       x Òs ôd èf Cy Xc       c ôd Ïf ìg Nx §v       v ½f Og ğh ™c ‚b       b äg äh j `v ±n       n ùh ôj k b üm       m Öj Ék 0l n         c ôv b dn       , èb èn èm       . èn èm             Øÿÿÿvk    ø§      SymbolsToInsert ğÿÿÿ' . -       ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   `§ `= ÿÿÿÿ        (          &   {31C04C26-5D28-4913-A7DF-0122B44927D1}  Øÿÿÿvk ´   ¨§      AccentTable     Hÿÿÿd e d é d è d ê d ë     d y d ÿ     d u d ù d û d ü     d i d ï d î     d o d ô d S    d a d à d â d æ     d s d ß d §     d c d ç d ©     d m d µ     d . d , d ? d : d ! d -     èÿÿÿ€§x§¨§Ø§ §Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk    ø§      KeyTableøüÿÿ  q Mw "a       w 
q je Ua Fs       e ¨w £r Ms ?d       r ‰e t ˆd zf       t ™r „y –f ag       y ®t _u ‚g Fh       u y Mi }h Xj       i Úu o .j Xk       o ui øp k pl       p ˆo ıl       a wq w Ès Oz       s Ow ne na ¬d az )x       d ue ôr s œf wz [x ıc       f År èt …d Æg Fx c èv       g t „y ~f Èh )c pv b       h ay Fu Ög j wv Jb ùn       j #u ôi jh ¾k Ob n Xm       k i o nj l ¯n #m       l Öo Fp •k äm       z Za Ùs ıd :x       x †s ıd èf xz Uc       c İd f ag {x ¶v       v #f 0g wh ;c §b       b g h Xj bv Ín       n ah j Fk mb ím       m „j Ék l dn         0c >v {b xn       , èb èn èm       . èn èm             Øÿÿÿvk    (§      SymbolsToInsert ğÿÿÿ' . -       ˆÿÿÿnk  ë›™ÕŞ¬Õ   0åG         ÿÿÿÿÿÿÿÿ   @§ `= ÿÿÿÿ        (   |      &   {3D2E093C-F00D-4BE8-B975-E5530BC5D28C}  Øÿÿÿvk `   Ø§      AccentTable     ˜ÿÿÿd 5d Q    d 3d ‘    d i d W    d Td M    d 8d K    d Ld J    d . d ? d : d ! d -         èÿÿÿ°§X§ˆ§¸§X#§Ğÿÿÿvk   €         HaveAccentTableScore    Ğÿÿÿvk   €         HaveKeyTableScore       àÿÿÿvk |  Ø§      KeyTable€ûÿÿ  9 F D i       F 9 C D i  2      C F : i  2 0      : C 5 2 0 ?      5 : = 0 ? @      = 5 3 ? @ >      3 = H @ > ;      H 3 I > ; 4      I H 7 ; 4 6    ect.defineProperty(globalObject, interfaceName, {
    configurable: true,
    writable: true,
    value: HTMLSelectElement
  });
};

const proxyHandlerCache = new WeakMap();
class ProxyHandler {
  constructor(globalObject) {
    this._globalObject = globalObject;
  }

  get(target, P, receiver) {
    if (typeof P === "symbol") {
      return Reflect.get(target, P, receiver);
    }
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc === undefined) {
      const parent = Object.getPrototypeOf(target);
      if (parent === null) {
        return undefined;
      }
      return Reflect.get(target, P, receiver);
    }
    if (!desc.get && !desc.set) {
      return desc.value;
    }
    const getter = desc.get;
    if (getter === undefined) {
      return undefined;
    }
    return Reflect.apply(getter, receiver, []);
  }

  has(target, P) {
    if (typeof P === "symbol") {
      return Reflect.has(target, P);
    }
    const desc = this.getOwnPropertyDescriptor(target, P);
    if (desc !== undefined) {
      return true;
    }
    const parent = Object.getPrototypeOf(target);
    if (parent !== null) {
      return Reflect.has(parent, P);
    }
    return false;
  }

  ownKeys(target) {
    const keys = new Set();

    for (const key of target[implSymbol][utils.supportedPropertyIndices]) {
      keys.add(`${key}`);
    }

    for (const key of Reflect.ownKeys(target)) {
      keys.add(key);
    }
    return [...keys];
  }

  getOwnPropertyDescriptor(target, P) {
    if (typeof P === "symbol") {
      return Reflect.getOwnPropertyDescriptor(target, P);
    }
    let ignoreNamedProps = false;

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      const indexedValue = target[implSymbol].item(index);
      if (indexedValue !== null) {
        return {
          writable: true,
          enumerable: true,
          configurable: true,
          value: utils.tryWrapperForImpl(indexedValue)
        };
      }
      ignoreNamedProps = true;
    }

    return Reflect.getOwnPropertyDescriptor(target, P);
  }

  set(target, P, V, receiver) {
    if (typeof P === "symbol") {
      return Reflect.set(target, P, V, receiver);
    }
    // The `receiver` argument refers to the Proxy exotic object or an object
    // that inherits from it, whereas `target` refers to the Proxy target:
    if (target[implSymbol][utils.wrapperSymbol] === receiver) {
      const globalObject = this._globalObject;

      if (utils.isArrayIndexPropName(P)) {
        const index = P >>> 0;
        let indexedValue = V;

        if (indexedValue === null || indexedValue === undefined) {
          indexedValue = null;
        } else {
          indexedValue = HTMLOptionElement.convert(indexedValue, {
            context: "Failed to set the " + index + " property on 'HTMLSelectElement': The provided value"
          });
        }

        ceReactionsPreSteps_helpers_custom_elements(globalObject);
        try {
          const creating = !(target[implSymbol].item(index) !== null);
          if (creating) {
            target[implSymbol][utils.indexedSetNew](index, indexedValue);
          } else {
            target[implSymbol][utils.indexedSetExisting](index, indexedValue);
          }
        } finally {
          ceReactionsPostSteps_helpers_custom_elements(globalObject);
        }

        return true;
      }
    }
    let ownDesc;

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      const indexedValue = target[implSymbol].item(index);
      if (indexedValue !== null) {
        ownDesc = {
          writable: true,
          enumerable: true,
          configurable: true,
          value: utils.tryWrapperForImpl(indexedValue)
        };
      }
    }

    if (ownDesc === undefined) {
      ownDesc = Reflect.getOwnPropertyDescriptor(target, P);
    }
    if (ownDesc === undefined) {
      const parent = Reflect.getPrototypeOf(target);
      if (parent !== null) {
        return Reflect.set(parent, P, V, receiver);
      }
      ownDesc = { writable: true, enumerable: true, configurable: true, value: undefined };
    }
    if (!ownDesc.writable) {
      return false;
    }
    if (!utils.isObject(receiver)) {
      return false;
    }
    const existingDesc = Reflect.getOwnPropertyDescriptor(receiver, P);
    let valueDesc;
    if (existingDesc !== undefined) {
      if (existingDesc.get || existingDesc.set) {
        return false;
      }
      if (!existingDesc.writable) {
        return false;
      }
      valueDesc = { value: V };
    } else {
      valueDesc = { writable: true, enumerable: true, configurable: true, value: V };
    }
    return Reflect.defineProperty(receiver, P, valueDesc);
  }

  defineProperty(target, P, desc) {
    if (typeof P === "symbol") {
      return Reflect.defineProperty(target, P, desc);
    }

    const globalObject = this._globalObject;

    if (utils.isArrayIndexPropName(P)) {
      if (desc.get || desc.set) {
        return false;
      }

      const index = P >>> 0;
      let indexedValue = desc.value;

      if (indexedValue === null || indexedValue === undefined) {
        indexedValue = null;
      } else {
        indexedValue = HTMLOptionElement.convert(indexedValue, {
          context: "Failed to set the " + index + " property on 'HTMLSelectElement': The provided value"
        });
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const creating = !(target[implSymbol].item(index) !== null);
        if (creating) {
          target[implSymbol][utils.indexedSetNew](index, indexedValue);
        } else {
          target[implSymbol][utils.indexedSetExisting](index, indexedValue);
        }
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }

      return true;
    }

    return Reflect.defineProperty(target, P, desc);
  }

  deleteProperty(target, P) {
    if (typeof P === "symbol") {
      return Reflect.deleteProperty(target, P);
    }

    const globalObject = this._globalObject;

    if (utils.isArrayIndexPropName(P)) {
      const index = P >>> 0;
      return !(target[implSymbol].item(index) !== null);
    }

    return Reflect.deleteProperty(target, P);
  }

  preventExtensions() {
    return false;
  }
}

const Impl = require("../nodes/HTMLSelectElement-impl.js");
