// jsdom/lib/jsdom/living/generated/HTMLVideoElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const HTMLConstructor_helpers_html_constructor = require("../helpers/html-constructor.js").HTMLConstructor;
const parseNonNegativeInteger_helpers_strings = require("../helpers/strings.js").parseNonNegativeInteger;
const ceReactionsPreSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPreSteps;
const ceReactionsPostSteps_helpers_custom_elements = require("../helpers/custom-elements.js").ceReactionsPostSteps;
const parseURLToResultingURLRecord_helpers_document_base_url =
  require("../helpers/document-base-url.js").parseURLToResultingURLRecord;
const serializeURLwhatwg_url = require("whatwg-url").serializeURL;
const implSymbol = utils.implSymbol;
const ctorRegistrySymbol = utils.ctorRegistrySymbol;
const HTMLMediaElement = require("./HTMLMediaElement.js");

const interfaceName = "HTMLVideoElement";

exports.is = value => {
  return utils.isObject(value) && utils.hasOwn(value, implSymbol) && value[implSymbol] instanceof Impl.implementation;
};
exports.isImpl = value => {
  return utils.isObject(value) && value instanceof Impl.implementation;
};
exports.convert = (value, { context = "The provided value" } = {}) => {
  if (exports.is(value)) {
    return utils.implForWrapper(value);
  }
  throw new TypeError(`${context} is not of type 'HTMLVideoElement'.`);
};

function makeWrapper(globalObject) {
  if (globalObject[ctorRegistrySymbol] === undefined) {
    throw new Error("Internal error: invalid global object");
  }

  const ctor = globalObject[ctorRegistrySymbol]["HTMLVideoElement"];
  if (ctor === undefined) {
    throw new Error("Internal error: constructor HTMLVideoElement is not installed on the passed global object");
  }

  return Object.create(ctor.prototype);
}

exports.create = (globalObject, constructorArgs, privateData) => {
  const wrapper = makeWrapper(globalObject);
  return exports.setup(wrapper, globalObject, constructorArgs, privateData);
};

exports.createImpl = (globalObject, constructorArgs, privateData) => {
  const wrapper = exports.create(globalObject, constructorArgs, privateData);
  return utils.implForWrapper(wrapper);
};

exports._internalSetup = (wrapper, globalObject) => {
  HTMLMediaElement._internalSetup(wrapper, globalObject);
};

exports.setup = (wrapper, globalObject, constructorArgs = [], privateData = {}) => {
  privateData.wrapper = wrapper;

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: new Impl.implementation(globalObject, constructorArgs, privateData),
    configurable: true
  });

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }
  return wrapper;
};

exports.new = globalObject => {
  const wrapper = makeWrapper(globalObject);

  exports._internalSetup(wrapper, globalObject);
  Object.defineProperty(wrapper, implSymbol, {
    value: Object.create(Impl.implementation.prototype),
    configurable: true
  });

  wrapper[implSymbol][utils.wrapperSymbol] = wrapper;
  if (Impl.init) {
    Impl.init(wrapper[implSymbol]);
  }
  return wrapper[implSymbol];
};

const exposed = new Set(["Window"]);

exports.install = (globalObject, globalNames) => {
  if (!globalNames.some(globalName => exposed.has(globalName))) {
    return;
  }

  if (globalObject.HTMLMediaElement === undefined) {
    throw new Error("Internal error: attempting to evaluate HTMLVideoElement before HTMLMediaElement");
  }
  class HTMLVideoElement extends globalObject.HTMLMediaElement {
    constructor() {
      return HTMLConstructor_helpers_html_constructor(globalObject, interfaceName, new.target);
    }

    get width() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get width' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        let value = esValue[implSymbol].getAttributeNS(null, "width");
        if (value === null) {
          return 0;
        }
        value = parseNonNegativeInteger_helpers_strings(value);
        return value !== null && value >= 0 && value <= 2147483647 ? value : 0;
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    set width(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set width' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      V = conversions["unsigned long"](V, {
        context: "Failed to set the 'width' property on 'HTMLVideoElement': The provided value"
      });

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const n = V <= 2147483647 ? V : 0;
        esValue[implSymbol].setAttributeNS(null, "width", String(n));
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    get height() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get height' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        let value = esValue[implSymbol].getAttributeNS(null, "height");
        if (value === null) {
          return 0;
        }
        value = parseNonNegativeInteger_helpers_strings(value);
        return value !== null && value >= 0 && value <= 2147483647 ? value : 0;
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    set height(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'set height' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      V = conversions["unsigned long"](V, {
        context: "Failed to set the 'height' property on 'HTMLVideoElement': The provided value"
      });

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const n = V <= 2147483647 ? V : 0;
        esValue[implSymbol].setAttributeNS(null, "height", String(n));
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    get videoWidth() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get videoWidth' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      return esValue[implSymbol]["videoWidth"];
    }

    get videoHeight() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get videoHeight' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      return esValue[implSymbol]["videoHeight"];
    }

    get poster() {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(esValue)) {
        throw new TypeError("'get poster' called on an object that is not a valid instance of HTMLVideoElement.");
      }

      ceReactionsPreSteps_helpers_custom_elements(globalObject);
      try {
        const value = esValue[implSymbol].getAttributeNS(null, "poster");
        if (value === null) {
          return "";
        }
        const urlRecord = parseURLToResultingURLRecord_helpers_document_base_url(
          value,
          esValue[implSymbol]._ownerDocument
        );
        if (urlRecord !== null) {
          return serializeURLwhatwg_url(urlRecord);
        }
        return conversions.USVString(value);
      } finally {
        ceReactionsPostSteps_helpers_custom_elements(globalObject);
      }
    }

    set poster(V) {
      const esValue = this !== null && this !== undefined ? this : globalObject;

      if (!exports.is(es      Formats àÿÿÿvk Û   ¸¡      Bitmap  hbin                         ğşÿÿS y s t e m . W e b . U I . M o b i l e C o n t r o l s . C u s t o m V a l i d a t o r ,   S y s t e m . W e b . M o b i l e ,   V e r s i o n = 4 . 0 . 0 . 0 ,   C u l t u r e = n e u t r a l ,   P u b l i c K e y T o k e n = b 0 3 f 5 f 7 f 1 1 d 5 0 a 3 a         Øÿÿÿvk Z   X¡      HelpKeyword      ÿÿÿS y s t e m . W e b . U I . M o b i l e C o n t r o l s . C u s t o m V a l i d a t o r      ÿÿÿ«   ì½`I–%&/mÊ{JõJ×àt¡€`$Ø@ìÁˆÍæ’ìiG#)«*ÊeVe]f@Ìí¼÷Ş{ï½÷Ş{ï½÷º;N'÷ßÿ?\fdlöÎJÚÉ!€ªÈ?~|?"~ÿû×ø5~hÿÿ5ş°›Ûlz×ü?òıÿıÿßƒÿŞ¥ÿÿaÄ&Ÿğ_ãŸø'ÿ‰Áÿã=óÿ½¿	‡.>Àáë¾oñø€÷
ô~§ÿ÷yzïàÿÃ–ÿßæ½oêÿ±ş˜x|İ¾ÿŸ   ÿÿ Øÿÿÿvk   € ÿ        TransparentColorĞÿÿÿvk   €         MinimumFrameworkVersion àÿÿÿvk   €Ğ~       Index   Øÿÿÿvk ª   8£      AssemblyName    PÿÿÿS y s t e m . W e b . M o b i l e ,   V e r s i o n = 4 . 0 . 0 . 0 ,   C u l t u r e = n e u t r a l ,   P u b l i c K e y T o k e n = b 0 3 f 5 f 7 f 1 1 d 5 0 a 3 a     àÿÿÿvk Z   ¤    TypeName ÿÿÿS y s t e m . W e b . U I . M o b i l e C o n t r o l s . C u s t o m V a l i d a t o r     àÿÿÿvk Ø   ˆ¤    TFMs°+ ÿÿÿ. N E T F r a m e w o r k , V e r s i o n = v 2 . 0 ; . N E T F r a m e w o r k , V e r s i o n = v 3 . 0 ; . N E T F r a m e w o r k , V e r s i o n = v 3 . 5 ; . N E T F r a m e w o r k , V e r s i o n = v 4 . 0       Ğÿÿÿvk N   ˜¥      TargetedItemProvider    ¨ÿÿÿ{ E 7 F 8 5 1 C 8 - 6 2 6 7 - 4 7 9 4 - B 0 F E - 7 B C A B 6 D A C B B 4 }         ¸ÿÿÿvk)   €         UseProjectTargetFrameworkVersionInTooltip        ÿÿÿnk  LV»„ŸÚ   ¸        ÿÿÿÿÿÿÿÿ   8§`’= ÿÿÿÿ        @   ‹         ItemContent     Ğÿÿÿvk j   È¦    .NET Toolbox Item+ ,ÿÿÿ–§ı;pC¦yVk²ˆû    ÿÿÿÿ          BObsolete: use ToolboxService to retrieve toolbox item information.  àÿÿÿ˜¦X§¨© ª¨ª    Øÿÿÿvk   €§      CF_NDP_TYPENAME ğşÿÿÿşS y s t e m . W e b . U I . M o b i l e C o n t r o l s . C u s t o m V a l i d a t o r ,   S y s t e m . W e b . M o b i l e ,   V e r s i o n = 4 . 0 . 0 . 0 ,   C u l t u r e = n e u t r a l ,   P u b l i c K e y T o k e n = b 0 3 f 5 f 7 f 1 1 d 5 0 a 3 a       Øÿÿÿvk L   ¸¨      CF_CACHEKEY     °ÿÿÿ–§ı;pC¦yVk²ˆû    ÿÿÿÿ          $2ed2d9fc-e4df-4af5-8fca-19af9dfa7b58Ğÿÿÿvk à   8©      CF_TOOLBOXITEMCONTAINER ÿÿÿ ,  ì½`I–%&/mÊ{JõJ×àt¡€`$Ø@ìÁˆÍæ’ìiG#)«*ÊeVe]f@Ìí¼÷Ş{ï½÷Ş{ï½÷º;N'÷ßÿ?\fdlöÎJÚÉ!€ªÈ?~|?"şŒ¿ò/ø?ËÃÕÉ_qı“¿ÙÛ¿åúÕ¿Æ¯ùkü¿ÆÿM~âùMùúç×ú5_û×ø
